src/storage/txn/process.rs

```diff
@@ -642,7 +642,9 @@ fn process_write_impl<S: Snapshot, L: LockMgr>(
             let key_hashes = gen_key_hashes_if_needed(&lock_mgr, &keys);
 
             let mut txn = MvccTxn::new(snapshot, start_ts, !ctx.get_not_fill_cache())?;
-            let is_pessimistic_txn = txn.cleanup(keys.pop().unwrap(), current_ts)?;
+            // If the primary lock of a pessimistic transaction is cleaned up, it must be protected
```

Comment:
How do you know it is primary lock?

Reply:
In a typical ti-client implementation, cleanup is only used for primary keys.

Reply:
Don't depend on these hidden rules, please add a new API like rollback_primary to do such thing.

Reply:
There are no _public_ rules. So there are no so-called _hidden_ rules. I believe cleanup is designed to rollback primary and resolve lock, otherwise it doesn't need to return the commit ts. The problem is that the API name is terrible. I don't know why it's called this name.